extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-widget.tsx

```diff
@@ -0,0 +1,481 @@
+        this.update();
+
+        await this.chePluginManager.installVSCodeExtension(extension);
+        await this.clearFilter();
+        return true;
+    }
+
+    protected render(): React.ReactNode {
+        const chePluginListControls = <ChePluginListControls
+            chePluginMenu={this.chePluginMenu}
+            filter={this.currentFilter}
+            update={this.onChangeFilter}
+            status={this.status} />;
+
+        return <React.Fragment>
+            {this.renderUpdateWorkspaceNotification()}
+            {chePluginListControls}
+            {this.renderPluginList()}
+        </React.Fragment>;
+    }
+
+    // Restart your workspace to apply changes
+    protected renderUpdateWorkspaceNotification(): React.ReactNode {
+        if (this.needToRestartWorkspace) {
+            const notificationStyle = this.hidingRestartWorkspaceNotification ? 'notification hiding' : 'notification';
+            return <div className='che-plugins-notification' >
+                <div className={notificationStyle}>
+                    <div className='notification-message' onClick={this.restartWorkspace}>
+                        <i className='fa fa-check-circle'></i>&nbsp;
+                        Click here to apply changes and restart your workspace
+                    </div>
+
+                    <div className='notification-control'>
+                        <div className='notification-hide' onClick={this.hideNotification}>
+                            <i className='fa fa-close alert-close' ></i>
+                        </div>
+                    </div>
+                </div>
+            </div>;
+        }
+
+        return undefined;
+    }
+
+    protected renderPluginList(): React.ReactNode {
+        // STATUS: loading
+        if (this.status === 'loading') {
+            return <div className='spinnerContainer'>
+                <div className='fa fa-spinner fa-pulse fa-3x fa-fw'></div>
+            </div>;
+        }
+
+        // STATUS: failed
+        if (this.status === 'failed') {
+            return <AlertMessage type='ERROR' header='Your registry is invalid' />;
+        }
+
+        // STATUS: ready
+        if (!this.plugins.length) {
+            return <AlertMessage type='INFO' header='No plugins currently available' />;
+        }
+
+        const list = this.plugins.map(plugin =>
+            <ChePlugin key={plugin.key}
+                plugin={plugin} pluginManager={this.chePluginManager}></ChePlugin>);
+        return <div className='che-plugin-list'>
+            {list}
+        </div>;
+    }
+
+    protected restartWorkspace = async e => {
+        await this.chePluginManager.restartWorkspace();
+    }
+
+    protected hideNotification = async e => {
+        this.hidingRestartWorkspaceNotification = true;
+        this.update();
+
+        setTimeout(() => {
+            this.needToRestartWorkspace = false;
+            this.hidingRestartWorkspaceNotification = false;
+            this.update();
```

Comment:
Why do we need two update in this method?

Reply:
I'm playing with the animation here.
See the whole method.
```
    protected hideNotification = async e => {
        this.hidingRestartWorkspaceNotification = true;
        this.update();

        setTimeout(() => {
            this.needToRestartWorkspace = false;
            this.hidingRestartWorkspaceNotification = false;
            this.update();
        }, 500);
    }
```